File: web/e2_viewer/src/document/document.ts
/** The Document/Store/Model system manages te primary application state.
 * 
 * The core concepts are Document, Store, Model, and Action.
 * 
 * A Document is the top-level container, owner and manager of application state. All state changes are initiated via a Document.
 *
 * Stores lie between Documents and Models. Stores protect Models from direct mutation, and provide controlled access for changing Model state. In future,
 * stores may also provide additional services such as undo/redo, state change logging, and asynchronous state changes.
 * 
 * Models represent the actual application state. Note that Document and Store are not intended to be subclassed. All of the specialization for
 * particular applications is handled by Model subclasses.
 *
 * An Action is a {type, payload} pair that specifies the operation to be performed and the payload data for that operation.
 * The operations (so-called "action functions") are registered with the Document during application initialization.
 * 
 * To initiate a state change, a client dispatches an Action to the Document, which looks up the corresponding action function and
 * invokes it with the supplied payload. The action function then uses the Store's changeState() method to modify the Model state in a controlled way -
 * specifically, the action function does not have direct write access to the Model - instead it must modify the Model inside a callback passed to it by the store's
 * changeState() method.
 *
 * Implementation notes:
 * TODO: add an isMutable flag to Store to prevent changeState calls outside of the callback.
 * TODO: add an "active action" concept to Document to track the currently executing action and prevent state changes outside of an action context.
 */

import { Store } from './store'; // Adjust import path as needed

export interface ActionSpec {
    type: string;
    payload: any;
}

export interface ActionDef {
    type: string;
    function: (document: Document, payload: any) => void;
}

export class Document {
    private stores: Map<string, Store> = new Map();
    private actionFunctions: Map<string, (document: Document, payload: any) => void> = new Map();

    constructor(stores?: Map<string, Store>) {
        if (stores) {
            this.stores = stores;
        }
    }

    storeAt(key: string): Store | undefined {
        return this.stores.get(key);
    }

    registerActionFunction(action: ActionDef): void {
        this.actionFunctions.set(action.type, action.function);
    }

    dispatchAction(action: ActionSpec): boolean {
        const actionFunction = this.actionFunctions.get(action.type);
        if (actionFunction) {
            actionFunction(this, action.payload);
            return true;
        }
        return false;
    }

    toString(): string {
        // Implement string representation as needed
        return `Document with ${this.stores.size} stores`;
    }
}

File: engines/src/document/document.cpp
#include <map>
#include <string>
#include <functional>
#include <iostream>
#include <nlohmann/json.hpp>

#include "document/document.h"

using json = nlohmann::json;

/**
 * The Document class holds the primary application state.
 * The primary concepts are Document, Store, Model, and Action.
 * A Document contains one or more Stores. 
 * Each Store contains a Model (the data).
 * State changes to the Model are made via Actions.
 * An application is expected to have a single Document instance. 
 * Some applications will employ just a single Store and Model within their single Document to store all of their state.
 * Other applications may choose to have multiple Stores and Models, each holding a different aspect of the state.
 * Client code subclasses Model to implement application-specific data structures.
 * Documents and Stores are not subclassed. They are designed to be generic, able to hold any type of Model.
 * Stores are concerned with lifecycle management of Models - things like undo/redo, 
 * persistence, locking, logging etc. (Most of these are not implemented yet).
 * Document provides the orchestration of Actions - registering action functions, dispatching actions to 
 * the appropriate function etc.
 * So, how does a client make a state change?
 * Firstly, note that all state changes are made in so-called "action functions". 
 * Action functions are regular functions, written by the application developer and registered
 * with the document at application startup time.
 * To invoke an action function, the client dispatches an Action to the Document. An Action is simply a {type, payload} pair.
 * The type is a string that identifies the action function to be invoked. The payload is a JSON object that contains any
 * parameters needed by the action function.
 * The Document looks up the action function registered for that action type, and invokes it with the payload.
 * Each action function then makes state changes to the Model via a state change callback on the Store that contains the Model.
 * It is this callback mechanism that enables Stores to manage lifecycle events, and to prevent uncontrolled state changes to Models.
 * As a convenience, the document also provides a post-state-change callback that gets invoked after each state change.
 * This can be used to trigger UI updates or other operations that are common to all state changes in an application.
 * Actions are chainable (one action function can dispatch another action function via the appropriate interface on the Document)
 * Actions are replayable (the Document could log all actions to a file, and then replay them later to restore state)
 * Actions are simple (just JSON data - no function pointers or other complexities)
 * Action can be dispatched across process boundaries by serializing to JSON text, and sending via stdin/stdout, sockets etc.
 * There is a class called DocumentService that provides a simple stdin/stdout IPC mechanism for dispatching actions to a Document 
 * and receiving actions in return.
 */

namespace e2 {

    Document::~Document() {
        for (auto& pair : m_stores) {
            delete pair.second;
        }
    }

    Store& Document::storeAt(const std::string& key) const {
        return *m_stores.at(key);
    }

    void Document::registerActionFunction(const ActionDef& action) {
        m_actionFunctions[action.type] = action.function;
    }

    ActionResult Document::dispatchAction(const ActionSpec& action) {
        try {
            //std::cerr << "Dispatching action of type: " << action.type << " with payload: " << action.payload.dump() << std::endl;  //--- DEBUG ---
            auto it = m_actionFunctions.find(action.type);
            if (it == m_actionFunctions.end()) {
                // std::cerr << "Unknown action type: " << action.type << std::endl;   <--- DEBUG ---
                return ActionResult::UNKNOWN_ACTION;
            }
            auto actionFunction = it->second;
            actionFunction(*this, action.payload);
        } catch (const json::exception& e) {
            std::cerr << "Error dispatching action: " << action.type << " - " << e.what() << std::endl;  // <--- DEBUG ---
            return ActionResult::INVALID_PAYLOAD;
        } catch (const std::exception& e) {
            // std::cerr << "Error dispatching action: " << action.type << " - " << e.what() << std::endl; // <--- DEBUG ---
            return ActionResult::INTERNAL_ERROR;
        }
        return ActionResult::SUCCESS;
    }

    void Document::dispatchClientAction(const ActionSpec& action) {
        m_clientActions.push_back(action);
    }

    const std::vector<ActionSpec>& Document::getClientActions() const {
        return m_clientActions;
    }

    void Document::clearClientActions() {
        m_clientActions.clear();
    }

    std::ostream& operator<<(std::ostream& os, const Document& doc) {
        os << "Document with " << doc.m_stores.size() << " stores." << std::endl;
        for (const auto& pair : doc.m_stores) {
            os << "Store key: " << pair.first << std::endl;
            os << *(pair.second) << std::endl;
        }
        return os;
    }   
};  